File: app/src/main/java/pl/laskowski/marcin/solving/deduction/algorithm/HintEliminationAlgorithm.java
package pl.laskowski.marcin.solving.deduction.algorithm;

import pl.laskowski.marcin.model.Field;
import pl.laskowski.marcin.model.Region;
import pl.laskowski.marcin.model.SudokuHintGrid;

import java.util.List;
import java.util.Set;

/**
 * Created by Marcin Laskowski.
 */

public class HintEliminationAlgorithm extends DeductionAlgorithm {

    public static class Factory implements DeductionAlgorithm.Factory {

        public HintEliminationAlgorithm instance(List<Region> regions, SudokuHintGrid sudokuHintGrid) {
            return new HintEliminationAlgorithm(regions, sudokuHintGrid);
        }

    }

    private HintEliminationAlgorithm(List<Region> regions, SudokuHintGrid possibilities) {
        super(regions, possibilities);
    }

    @Override
    protected void solve(Region region) {
        for (Field field : region.fullFields()) {
            ensureFullFieldHaveNoHints(field);
            removeHintFromRegion(region, field.value());
        }
    }

    private void ensureFullFieldHaveNoHints(Field field) {
        if (!possibilities.isEmpty(field)) {
            possibilities.clear(field);
            notifyChange();
        }
    }

    private void removeHintFromRegion(Region region, int value) {
        for (Field field : region) {
            if (possibilities.remove(field, value)) {
                notifyChange();
            }
        }
    }

}

File: src/main/kotlin/io/github/ilikeyourhat/kudoku/solving/deduction/algorithm/HintEliminationAlgorithm.kt
package io.github.ilikeyourhat.kudoku.solving.deduction.algorithm

import io.github.ilikeyourhat.kudoku.model.Cell
import io.github.ilikeyourhat.kudoku.model.Region
import io.github.ilikeyourhat.kudoku.model.hint.SudokuHintGrid

class HintEliminationAlgorithm(
    regions: List<Region>,
    possibilities: SudokuHintGrid
) : DeductionAlgorithm(regions, possibilities) {

    class Factory : DeductionAlgorithm.Factory {
        override fun instance(regions: List<Region>, possibilities: SudokuHintGrid): HintEliminationAlgorithm {
            return HintEliminationAlgorithm(regions, possibilities)
        }
    }

    override fun solve(): Boolean {
        return regions
            .map { solve(it) }
            .any { it }
    }

    override fun solve(region: Region): Boolean {
        var changed = false
        for (cell in region.fullCells()) {
            changed = changed or
                ensureFullCellHaveNoHints(cell) or
                removeHintFromRegion(region, cell.value())
        }
        return changed
    }

    private fun ensureFullCellHaveNoHints(cell: Cell): Boolean {
        if (!possibilities.isEmpty(cell)) {
            possibilities.clear(cell)
            return true
        }
        return false
    }

    private fun removeHintFromRegion(region: Region, value: Int): Boolean {
        var changed = false
        for (cell in region) {
            changed = changed or possibilities.remove(cell, value)
        }
        return changed
    }
}
